faster_whisper/transcribe.py

```diff
@@ -82,14 +101,587 @@ class TranscriptionInfo(NamedTuple):
+
+        segments = self._batched_segments_generator(
+            audio,
+            vad_segments,
+            sampling_rate,
+            batch_size,
+            batched_options,
+            log_progress,
+        )
+
+        return segments, info
+
+    def _batched_segments_generator(
+        self, audio, vad_segments, sampling_rate, batch_size, options, log_progress
+    ):
+        seg_idx = 0
+        total_segments = len(vad_segments)
+        for idx, out in enumerate(
+            self.__call__(
+                self.audio_split(audio, vad_segments, sampling_rate),
+                batch_size=batch_size,
+                options=options,
+            )
+        ):
+            if log_progress:
+                percent_complete = ((idx + 1) / total_segments) * 100
+                self.model.logger.info(f"Progress: {percent_complete:.2f}%...")
+
+            responses = out["output"]
+            if batch_size == 1:
+                responses = responses[0]
+
+            for response in responses:
+                seg_idx += 1
+                segments = Segment(
+                    seek=int(responses[-1]["end"] * self.model.frames_per_second),
+                    id=seg_idx,
+                    text=response["text"],
+                    start=round(response["start"], 3),
+                    end=round(response["end"], 3),
+                    words=(None if not options.word_timestamps else response["words"]),
+                    tokens=response["tokens"],
+                    avg_logprob=response["avg_logprob"],
+                    no_speech_prob=response["no_speech_prob"],
+                    compression_ratio=response["compression_ratio"],
+                )
+                yield segments
+
+        # revert the tokenizer if multilingual inference is enabled
+        if self.preset_language is None:
+            self.tokenizer = None
+        self.last_speech_timestamp = 0.0
+
+    def detect_language(self, audio: torch.Tensor):
+        to_cpu = (
+            self.model.model.device == "cuda" and len(self.model.model.device_index) > 1
+        )
+        segment = self.model.feature_extractor(audio, padding=True, to_cpu=to_cpu)[
+            :, : self.model.feature_extractor.nb_max_frames
+        ]
+        encoder_output = self.model.encode(segment)
+        results = self.model.model.detect_language(encoder_output)
+        language_token, language_probability = results[0][0]
+        language = language_token[2:-2]
+        self.model.logger.info(
+            f"Detected language: {language} ({language_probability:.2f}) in first 30s of audio..."
+        )
+        all_language_probs = [(token[2:-2], prob) for (token, prob) in results[0]]
+        return language, language_probability, all_language_probs
+
+    def detect_language_multi_segment(
+        self, audio: Union[str, BinaryIO, torch.Tensor], params: Optional[dict] = None
+    ):
+        return self.model.detect_language_multi_segment(audio, params)
```

Comment:
I see no benefit of having this function, users can always call `pipeline.model.detect_language_multi_segment(audio, params)`
just for reducing functionality duplication, we should move all the functions that generalize to both sequential and batched inference to `WhisperModel`